timer/base_timer.go

```diff
@@ -0,0 +1,127 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+package timer
+
+import (
+	"errors"
+	"time"
+)
+
+// baseTimer is the type used for all the timers that won't have children
+// It's implementation is more lightweight than nodeTimer and can be used as a standalone timer using NewTimer
+type baseTimer struct {
+	timeCache
+
+	// start is the time when the timer was started
+	start time.Time
+
+	// spent is the time spent on the timer, set after calling stop
+	spent time.Duration
+
+	// config is the configuration of the timer
+	config Config
+
+	// parentComponent is the component stored in the map of the parent timer. We keep a reference to it to update its spent time
+	// once the timer is stopped. It may be nil if the timer is not a child of a tree timer, like when calling NewTimer.
+	parentComponent *component
+
+	// parent is the parent timer. It is used to progate the stop of the timer to the parent timer and get the remaining time in case the budget has to be inherited.
+	parent NodeTimer
+}
+
+var _ Timer = (*baseTimer)(nil)
+
+// NewTimer creates a new Timer with the given options. You have to specify either the option WithBudget or WithUnlimitedBudget.
+func NewTimer(options ...Option) (Timer, error) {
```

Comment:
The issue I have with this is that this would give a proactive use to the `UnlimitedBudget` special value and it is quite like an implementation detail (that I had to export because of the behaviour of `Remaining()` but still). I would prefer to have `NewTimerWithUnlimitedBudget` function but, for the same reason as I don't want a `NewContextWithBudget`, I don't feel that this is futur proof and the Option pattern has strictly no performance impact so :shrug: